arb_os/retrybuffer.mini

```diff
@@ -0,0 +1,357 @@
+//
+// Copyright 2021, Offchain Labs, Inc. All rights reserved.
+//
+
+use accounts::getGlobalAccountStore;
+use accounts::setGlobalAccountStore;
+use accounts::accountStore_get;
+use accounts::accountStore_set;
+use accounts::account_addToEthBalance;
+use accounts::account_deductFromEthBalance;
+
+use evmCallStack::evmCallStack_callRetryableTxAndReturn;
+
+use evmlogs::EvmLogs;
+
+use gasAccounting::GasUsage;
+use gasAccounting::txFeeStats_zero;
+use gasAccounting::chargeForRetryableTx;
+use gasAccounting::getRetryableTxPrice;
+use gasAccounting::txFeeStats_zero;
+
+use inbox::IncomingRequest;
+use inbox::inbox_currentTimestamp;
+
+use messages::TxRequestData;
+
+use output::emitTxReceipt;
+
+use std::bytearray::ByteArray;
+use std::bytearray::bytearray_new;
+use std::bytearray::bytearray_size;
+use std::bytearray::bytearray_get256;
+use std::bytearray::bytearray_set256;
+use std::bytearray::bytearray_extract;
+
+use std::keccak::keccak256;
+
+use std::queue::Queue;
+use std::queue::queue_new;
+use std::queue::queue_put;
+use std::queue::queue_get;
+
+use core::kvs::Kvs;
+use core::kvs::builtin_kvsDelete;
+
+
+type RetryableTx = struct {
+    txId: uint,
+    sender: address,
+    destination: address,
+    callvalue: uint,
+    beneficiary: address,
+    calldata: ByteArray,
+    expiryTime: uint,
+    l1Message: IncomingRequest,
+}
+
+type RetryQueueItem = struct {
+    txId: uint,         // id of item to check
+    checkTime: uint,    // when to check for expiration of the item
+}
+
+// RetryBuffer holds retryable transactions.
+// byTxId holds all active RetryableTx's, indexed by their txId
+// fifo holds RetryQueueItems, which direct us to check for expiration of a particular item at a particular time.
+// Every time an RetryableTx's lifetime is renewed, a fresh fifo item is created,
+//          with checkTime equal to the RetryableTx's new expiry time.
+var retryBuffer: struct {
+    byTxId: map<uint, RetryableTx>,
+    fifo: Queue,     // not in strict expirationTime order, but this doesn't affect correctness
+    txLifetimeSeconds: uint
+};
+
+public impure func retryBuffer_init() {
+    retryBuffer = struct {
+        byTxId: newmap<uint, RetryableTx>,
+        fifo: queue_new(),
+        txLifetimeSeconds: const::RetryBuffer_DefaultLifetimeSeconds,
+    };
+}
+
+public impure func retryBuffer_getLifetime() -> uint {
+    let _ = retryBuffer_trimOne();
```

Comment:
I think we might want to just trim on otherwise mutating methods

Reply:
Makes sense.  Changed this.